runtime/common/src/main/kotlin/kotlinx/serialization/json/serializers/JsonElementSerializer.kt

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2017-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.json.serializers
+
+import kotlinx.serialization.*
+import kotlinx.serialization.internal.*
+import kotlinx.serialization.json.*
+
+@Serializer(forClass = JsonElement::class)
+object JsonElementSerializer : KSerializer<JsonElement> {
+    override fun serialize(encoder: Encoder, obj: JsonElement) {
+        when (obj) {
+            is JsonPrimitive -> JsonPrimitiveSerializer.serialize(encoder, obj)
+            is JsonObject -> JsonObjectSerializer.serialize(encoder, obj)
+            is JsonArray -> JsonArraySerializer.serialize(encoder, obj)
+        }
+    }
+
+    override fun deserialize(decoder: Decoder): JsonElement {
+        val input = decoder as? JsonInput ?: error("JsonElement is serializable only when used by Json")
+        return input.readTree()
+    }
+
+    override val descriptor: SerialDescriptor = object : SerialClassDescImpl("JsonElementSerializer") {
+        override val kind: SerialKind
+            get() = UnionKind.SEALED
+
+        init {
+            addElement("JsonObject")
+            addElement("JsonLiteral")
+            addElement("JsonArray")
+        }
+    }
+}
+
+private object JsonObjectSerializer : KSerializer<JsonObject> {
+
+    override val descriptor: SerialDescriptor = JsonObjectDescriptor
+
+    override fun serialize(encoder: Encoder, obj: JsonObject) {
+        val composite = encoder.beginCollection(descriptor, obj.size, StringSerializer, JsonElementSerializer)
```

Comment:
I assume you can use delegation to MapSerializer for serialize, deserialize and descriptor, like: 

```kotlin
override fun serialize(encoder: Encoder, obj: JsonObject) {
    LinkedHashMapSerializer(StringSerializer, JsonElementSerializer).serialize(obj.content)
}
```

And use delegation to `LinkedHashMapClassDesc` to override descriptor name.